feat(transport/cc): make slow start pluggable and add hystart++

omansfeld opened the pull request:
[fdc7597](https://github.com/mozilla/neqo/pull/3278/commits/fdc7597df7d5c45ba5e2a4b74dc7b056a42b6d35) introduces the scaffolding to make slow start configurable by introducing a `SlowStart` trait and using a generic parameter with that trait bound in the `ClassicCongestionControl` struct. It also rewrites the current slow start implementation in `classic_slow_start.rs`.

[df81bdb](https://github.com/mozilla/neqo/pull/3278/commits/df81bdb76966a5e4fcbaa6ec61c639ed608ca0d4) makes the slow start algorithm configurable at run-time via the CLI/ConnectionParams, similar to how it works for Cubic/NewReno.

As a reminder: [fdc7597..df81bdb](https://github.com/mozilla/neqo/pull/3278/changes/BASE..df81bdb76966a5e4fcbaa6ec61c639ed608ca0d4) had already had a review cycle before I started working on HyStart++ and made the PR a draft again.

[41e8c04](https://github.com/mozilla/neqo/pull/3278/commits/41e8c04577bb69aa7947e0d00fbde6159ca5becb) implements HyStart++ as per [RFC 9406](https://datatracker.ietf.org/doc/html/rfc9406) with extensive tests added in [86b496d](https://github.com/mozilla/neqo/pull/3278/commits/86b496d41fc2256c9cdda01ee4137ca413d32e28).

With the latest commit [ae92b97](https://github.com/mozilla/neqo/pull/3278/commits/ae92b975b27770deedb4ea87a2683eb896b8b95f) I re-architected the `SlowStart` trait and the order in which slow start functionality is called from the congestion controller. Hystart functionality is unchanged between that commit and the former implementation, this just improves architecture and makes it more extensible for future work. 

It now separates functionality into the correct order of operations for slow start (`check for exit` -> `do the exit` -> `apply growth if not exiting` -> `do congestion avoidance if not in slow start anymore`) which also makes it easier to add things like [SUSS](https://dl.acm.org/doi/10.1145/3651890.3672234) or [rapid-start](https://datatracker.ietf.org/doc/html/draft-kazuho-ccwg-rapid-start-00) that accelerate slow start growth (but not touch the exit heuristics) in the future.

To maybe make review easier, the main parts are:

- `neqo-transport/src/cc/hystart.rs` contains the HyStart++ implementation.
- `neqo-transport/src/cc/tests/hystart.rs` contains tests for the HyStart++ implementation.
- `neqo-transport/src/cc/classic_slow_start.rs` has the standard slow start implementation that was previously hardcoded into `classic_cc.rs`.
- `neqo-transport/src/cc/classic_cc.rs` has the core `on_packets_acked` flow that calls everything else and the trait definition for `SlowStart`. There is also a small change to `on_packet_sent` for round-tracking.
- ... and then there is a lot of changing function calls in tests, scaffolding, etc.

I think it might make it the easiest to look at `classic_cc.rs` first and then follow the code flow from there into the HyStart++ implementation. I'm not sure it makes a lot of sense to look at commits one-by-one because of the re-architecture in the latest commit.

I also have some qlogs from validating the implementation against `h3.speed.cloudflare.com` where one can see HyStart++ working as intended. See this qvis screenshot where it exits at the same window size that is later converged on during congestion avoidance:

<img width="1459" height="393" alt="hystart_textbook-exit" src="https://github.com/user-attachments/assets/96874640-6e0b-43d9-af2a-bc7639b226ee" />

omansfeld commented on neqo-transport/src/cc/tests/cubic.rs:
just lots of function call changes here

omansfeld commented on neqo-transport/src/cc/tests/new_reno.rs:
same here

larseggert commented on neqo-transport/src/cc/tests/cubic.rs:
```diff
@@ -44,8 +44,15 @@ const fn cwnd_after_loss_slow_start(cwnd: usize, mtu: usize) -> usize {
-) -> (ClassicCongestionControl<Cubic>, CongestionControlStats) {
-    let mut cc = ClassicCongestionControl::new(Cubic::default(), Pmtud::new(IP_ADDR, MTU));
+) -> (
+    ClassicCongestionControl<ClassicSlowStart, Cubic>,
```
```suggestion
    cc:Classic<ss:Classic, Cubic>,
```
The module is called `cc`, so the `ClassicCongestionControl` -> `cc::Classic` rename should be straightforward. In order to be able to use `ss`, we need a module. (Also wonder if `cc` and `ss` are to terse and we should find better module names.)

omansfeld replied:
I don't think `ss` should be it's own module. For me it's the same category as `Cubic` or `NewReno`. I think this just looks very weird because `ClassicSlowStart` is not the best name (because it includes "SlowStart"). 

If it was `ClassicCongestionControl<HyStart, Cubic>` this would look much more sensible imo. I'm not sure what a better name for the standard slow start is though. I'm reluctant towards just naming it `SlowStart` or `Classic` because that might get confusing.

As for `ClassicCongestionControl` I think we should refactor the structure there a bit anyways because right now we don't have anything else then `Classic` and are not planning to implement BBR afaik. So it seems a bit unnecessary to have the `CongestionControl` trait that is only implemented by `ClassicCongestionControl`. I think it could just be `CongestionControl` without an extra trait. But that is a seperate PR and I'd like to focus on getting the functionality in before doing that.

But yes, ideally this would become something like `CongestionController<HyStart, Cubic>` (or any other combination).

larseggert commented on neqo-transport/src/cc/classic_slow_start.rs:
```diff
@@ -0,0 +1,50 @@
+            exit_slow_start = true;
+        }
+
+        (cwnd_increase, unused_acked_bytes, exit_slow_start)
```
```suggestion
        let cwnd_increase = min(ssthresh - curr_cwnd, acked_bytes + new_acked);
        let unused_acked_bytes = acked_bytes + new_acked - cwnd_increase;

        // This doesn't look like it is necessary, but it can happen
        // after persistent congestion.
        let exit_slow_start = curr_cwnd + cwnd_increase == ssthresh;

        (cwnd_increase, unused_acked_bytes, exit_slow_start)
```

larseggert commented on neqo-transport/src/cc/classic_slow_start.rs:
```diff
@@ -0,0 +1,50 @@
+        ssthresh: usize,
+        acked_bytes: usize,
+        new_acked: usize,
+    ) -> (usize, usize, bool) {
```
Does this return triple warrant a type?

omansfeld commented:
> I suggest only merging here once we have more than one slow start algorithm. Each pull request should improve the state of `main`. Merging here now only adds a dead abstraction, i.e. `SlowStart`, only implemented once. (Similar to `trait CongestionControl` today.)

As per @mxinden's comment I made this PR a draft again and will also add HyStart++ as part of it. I might still ask for review on single commits along the way to have some incremental feedback.

omansfeld commented:
I added another commit to make slow start configurable via connection parameters and the CLI. Would welcome another review once you have some time @mxinden!

Also a question, just to make sure I understood the parameter code structure correctly:

`ConnectionParameters` is what actually sets the cc parameters for the sender and is also what we would set via prefs from Firefox for e.g. experimentation in [neqo_glue](https://searchfox.org/firefox-main/rev/14c08f0368ead8bfdddec62f43e0bb5c8fd61289/netwerk/socket/neqo_glue/src/lib.rs#432-444), correct?

https://github.com/mozilla/neqo/blob/adcd4bddddce0fb83bbf3f6c01c3c740f6edc680/neqo-transport/src/sender.rs#L36-L44

But `QuicParameters` is used to set CLI flags in which case it passes them on to `ConnectionParams` when a client/server is created (of course it also contains more than just connection parameters).

https://github.com/mozilla/neqo/blob/adcd4bddddce0fb83bbf3f6c01c3c740f6edc680/neqo-bin/src/lib.rs#L234-L243

But from Firefox we [directly set our `ConnectionParameters`](https://searchfox.org/firefox-main/rev/14c08f0368ead8bfdddec62f43e0bb5c8fd61289/netwerk/socket/neqo_glue/src/lib.rs#432,457,468), so any defaults we have in Neqo basically don't matter if the consumer is Firefox, correct?

mxinden reviewed (commented):
Changes here thus far look good to me.

larseggert commented on neqo-transport/src/cc/hystart.rs:
```diff
@@ -0,0 +1,307 @@
+
+use crate::{cc::classic_cc::SlowStart, packet, rtt::RttEstimate};
+
+#[derive(Debug, Default, Clone, Copy)]
```
`Copy` seems costly. Do you even need it? And do you need `Default`?

larseggert commented on neqo-transport/src/cc/hystart.rs:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+#[derive(Debug, Default)]
```
Do you even need `Default`?

omansfeld replied:
No, see above.

omansfeld commented on neqo-bin/src/lib.rs:
```diff
@@ -160,7 +166,8 @@ impl Default for QuicParameters {
                 .expect("this value will always be less than u64::MAX"),
-            congestion_control: CongestionControlAlgorithm::Cubic,
+            cc_algorithm: CongestionControlAlgorithm::Cubic,
+            ss_algorithm: SlowStartAlgorithm::Classic,
```
I think we can shorten both `SlowStart` and `CongestionControl` to omit the `-Algorithm`.

omansfeld commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -107,6 +107,37 @@ pub trait WindowAdjustment: Display + Debug {
+    /// Handle packets being acknowledged during initial slow start and do computations to decide if
+    /// initial slow start should be exited to congestion avoidance.
+    ///
+    /// Returns `exit_to_ca`.
```
It is the return value to decide if we want to exit to congestion avoidance (`ca`). I renamed it from the `exit_slow_start` I had initially because I felt `exit_slow_start` was ambiguous as the HyStart RFC also uses "exiting slow start" when referencing going from slow start to conservative slow start.

I don't have a strong opinion here. If you think `exit_to_ca` is not clear enough I can fall back to `exit_slow_start` or another suggestion.

mxinden commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -107,6 +107,37 @@ pub trait WindowAdjustment: Display + Debug {
+/// growth rate during slow start is.
+pub trait SlowStart: Display + Debug {
+    /// Handle tracking RTT rounds through having the next packet number that is to be sent out
+    /// passed in.
```
Not sure I understand. Would the below be correct, too?


```suggestion
    /// Enables a trait implementor to track RTT rounds via the next packet number that is to be sent out.
```

mxinden commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -107,6 +107,37 @@ pub trait WindowAdjustment: Display + Debug {
+    /// passed in.
+    fn on_packet_sent(&mut self, sent_pn: packet::Number);
+    /// Handle packets being acknowledged during initial slow start and do computations to decide if
+    /// initial slow start should be exited.
```
```suggestion
    /// Handle packets being acknowledged during initial slow start. Returns whether
    /// initial slow start should be exited.
```

Feel free to ignore.

mxinden commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -107,6 +107,37 @@ pub trait WindowAdjustment: Display + Debug {
+    /// One example is the reduced growth in Conservative Slow Start (CSS) from HyStart++.
+    ///
+    /// Returns `cwnd_increase`.
+    fn maybe_change_cwnd_increase(
```
Nit pick: `maybe_` implies returning an `Option`, yet an implementor that doesn't alter the cwnd (classic slow start) just returns `cwnd_increase`. Is that intuitive?

If you return an `Option` instead, you could add a default implementation here:

``` rust
fn maybe_change_cwnd_increase(
        &mut self,
        cwnd_increase: usize,
        max_datagram_size: usize,
    ) -> Option<usize> {
    None
}
```

and then not implement it in classic slow start at all.

mxinden commented on neqo-transport/src/cc/hystart.rs:
```diff
@@ -0,0 +1,301 @@
+        if !self.in_css()
+            && self.enough_samples()
+            && self.current.current_round_min_rtt != Duration::MAX
+            && self.current.last_round_min_rtt != Duration::MAX
```
`Duration::Max` here represents the no-value value, overloading the meaning of a `usize`. I suggest leveraging Rust's `Option` here. In other words, I suggest making `current_round_min_rtt` and `last_round_min_rtt` an `Option`, then check for `is_none`.

Feel free to ignore in case it significantly complicates the implementation.

mxinden commented on neqo-transport/src/cc/hystart.rs:
```diff
@@ -0,0 +1,301 @@
+
+    pub const CSS_ROUNDS: usize = 5;
+
+    pub const NON_PACED_L: usize = 8;
```
Do any other implementations deviate from the suggested constants in the RFC?

mxinden commented on neqo-transport/src/cc/classic_slow_start.rs:
```diff
@@ -0,0 +1,51 @@
+    fn on_packet_sent(&mut self, _sent_pn: packet::Number) {}
+
+    fn on_packets_acked(&mut self, _rtt_est: &RttEstimate, _largest_acked: packet::Number) -> bool {
+        // Standard slow start does not have any heuristic for exiting initial slow start. Always
```
Preference for `ClassicSlowStart`. Most intuitive to me.

mxinden commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -291,27 +339,64 @@ impl<T: WindowAdjustment> CongestionControl for ClassicCongestionControl<T> {
+                qdebug!("Exited slow start by algorithm");
+                let exit_cwnd = self
+                    .slow_start
+                    .on_slow_start_exit(self.current.congestion_window);
```
I would prefer not introducing an abstraction that is only needed in a follow-up pull request (e.g. here the SEARCH pull request).

> I guess it could fundamentally be a part of `on_packets_acked` in either case, but imo the code in `classic_cc::on_packets_acked` is more readable by having those two seperate blocks of A: deciding if to exit slow start, and B: exiting slow start by setting the `ssthresh`.

I prefer a single function, preventing misuse, e.g. calling `on_slow_start_exit` without calling `self.slow_start_on_packets_acked`. There more restrictive the interface, the less misuse it allows.

mxinden commented on neqo-transport/src/cc/classic_cc.rs:
```diff
@@ -107,6 +107,31 @@ pub trait WindowAdjustment: Display + Debug {
+    fn calc_cwnd_increase(&mut self, new_acked: usize, _max_datagram_size: usize) -> usize {
+        new_acked
+    }
+}
```
Given that we use a concrete `SlowStart` implementation instance both for the initial slow start as well as the slow start when re-entered through persistent congestion, should there be a `SlowStart::reset` function which is called before re-entering slow start? Otherwise e.g. `HyStart` would continue off of its state from the initial slow start.

omansfeld commented:
Updated again in https://github.com/mozilla/neqo/pull/3278/commits/89a516977e45430a9bf77f972e1b33bc44f20ae0. Addressed nits from @mxinden, added comment about the non-initial slow start decision and added `SlowStart::reset()` to erase state after persistent congestion. Also added another test that asserts this.
